6pm-4-csv tutor assignment upload

bzamora020 opened the pull request:
In this PR, there has been added the functionality to upload a CSV to add tutor assignments on a mass scale instead of adding tutor assignments one by one. If there has not been created a previous Tutor or Course, it will create them. If they have been created, it won't modify them. The same case with the tutor assignment, this is to avoid duplicates. 


NOTE: There is 3 lines not covered in the test coverage, due to this happened to a previous PR and it is pretty tedious to make the test for those, it will be necessary to create an issue to fix that in a PR. Once this PR is merged

WadeVaresio commented on src/main/java/edu/ucsb/ucsbcslas/controllers/TutorAssignmentController.java:
```diff
@@ -14,6 +14,18 @@
 import org.springframework.web.bind.annotation.DeleteMapping;
 import org.springframework.web.bind.annotation.RestController;
 
+// Imports from BZ when adding uploading CSV functionality to Tutor Assignment
```
We should remove this comment, we always use git blame to find out who wrote a line of code.

WadeVaresio commented on src/main/java/edu/ucsb/ucsbcslas/controllers/TutorAssignmentController.java:
```diff
@@ -221,55 +235,109 @@
+        logger.info(incomingTutorAssignment);
+        JSONObject ta = new JSONObject(incomingTutorAssignment);
+        TutorAssignment newAssignment = new TutorAssignment();
+        logger.info("ta: ", ta.toString());
```
I believe `ta.toString()` is uneccessary, you can just replace with `ta` and it will implicitly call toString.

Up to you if you want to keep the `toString` invocation.

bzamora020 replied:
I actually didn't write that code. I believe some other student was also working on something in TutorAssignment controller, and when I merged the files some of their code came in the pull request

WadeVaresio replied:
I see, no worries then.

WadeVaresio commented on src/main/java/edu/ucsb/ucsbcslas/controllers/TutorAssignmentController.java:
```diff
@@ -221,55 +235,109 @@
+        TutorAssignment newAssignment = new TutorAssignment();
+        logger.info("ta: ", ta.toString());
+        JSONObject cInfo = new JSONObject(ta.get("course").toString());
+        logger.info("cInfo: ", cInfo.toString());
```
I believe `cInfoo.toString()` is uneccessary, you can just replace with `ta` and it will implicitly call toString.

Up to you if you want to keep the `toString` invocation.

bzamora020 replied:
Is the same case with this code. I wrote the one that was after the code on line 286 where it says, "Controller to upload CSV for tutorAssignment Service" 

I was a bit confused on how it got included on my PR

WadeVaresio replied:
I see, no worries then.

WadeVaresio commented on src/main/java/edu/ucsb/ucsbcslas/controllers/TutorAssignmentController.java:
```diff
@@ -221,55 +235,109 @@
+                if (tutorAssignmentList.isEmpty()) {
+                    currentTutorAssignment = new TutorAssignment(currentCourse, currentTutor, i.getAssignmentType());
+                    tutorAssignmentRepository.save(currentTutorAssignment);
+                } else {
+                    currentTutorAssignment = tutorAssignmentList.get(0);
+                }
```
I think this else statement can be removed, as `currentTutorAssignment` is not used after

bzamora020 replied:
Good catch!

WadeVaresio commented on src/main/java/edu/ucsb/ucsbcslas/controllers/TutorAssignmentController.java:
```diff
@@ -221,55 +235,109 @@
+    public ResponseEntity<String> uploadCSV(@RequestParam("csv") MultipartFile csv,
+            @RequestHeader("Authorization") String authorization) throws IOException {
+        logger.info("Starting upload CSV");
+        AppUser user = authControllerAdvice.getUser(authorization);
```
We don't do anything with `user` there should be a check to see if `user` is of the correct role.

bzamora020 replied:
A regular user can't see the upload tutor assignment option, which is why I decided to don't have to handle that case

JacquelineMai replied:
Even if the link to this page might be hidden from a regular user, they may still be able to access it directly via the url if we don't enforce authorization here

bzamora020 replied:
I updated that. But there is a catch exception error test that I am not passing due to I changed some of the code (it was passing before I introduced an if statement) Is it possible to fix that on another  PR?

WadeVaresio replied:
Tests have been introduced to cover this case, marking conversation as resolved.

WadeVaresio commented on src/main/java/edu/ucsb/ucsbcslas/controllers/TutorController.java:
```diff
@@ -81,7 +81,8 @@
+    public ResponseEntity<String> uploadCSV(@RequestParam("csv") MultipartFile csv,
+            @RequestHeader("Authorization") String authorization) throws IOException {
         logger.info("Starting upload CSV");
         AppUser user = authControllerAdvice.getUser(authorization);
```
We don't do anything with `user` we should check whether the current user is authorized to make this POST request.

bzamora020 replied:
I did not work on the PR for uploading CSV's for the Tutor. I think the lines got modified when I auto formatted this file

JacquelineMai replied:
This is something that should definitely be noted in an issue, but I agree doesn't pertain directly to this PR

WadeVaresio commented on src/main/java/edu/ucsb/ucsbcslas/entities/TutorAssignment.java:
```diff
@@ -10,6 +10,8 @@
 import javax.persistence.JoinColumn;
 import javax.persistence.ManyToOne;
 
+import com.opencsv.bean.CsvBindByPosition;
```
Unused import, we should remove.

bzamora020 replied:
done

WadeVaresio commented on src/main/java/edu/ucsb/ucsbcslas/models/TutorAssignmentModel.java:
```diff
@@ -0,0 +1,165 @@
+import java.util.Objects;
+
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.GeneratedValue;
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+import com.opencsv.bean.CsvBindByName;
```
Unused imports, we should remove.

bzamora020 replied:
done

WadeVaresio commented on src/main/java/edu/ucsb/ucsbcslas/models/TutorAssignmentModel.java:
```diff
@@ -0,0 +1,165 @@
+
+public class TutorAssignmentModel {
+
+    @Column(nullable = false)
```
Since we are not directly storing this class in a repository, we can remove all Column annotations

bzamora020 replied:
Done

WadeVaresio commented on src/test/java/edu/ucsb/ucsbcslas/controllers/TutorAssignmentControllerTests.java:
```diff
@@ -10,6 +10,22 @@
 import org.springframework.test.web.servlet.MockMvc;
 import org.springframework.test.web.servlet.MvcResult;
 
+// Imports from BZ when implementing uploading CSV functionality to Tutor Assignment
```
We should remove this line, as we can always use git blame

bzamora020 replied:
Done

WadeVaresio commented on src/test/java/edu/ucsb/ucsbcslas/controllers/TutorAssignmentControllerTests.java:
```diff
@@ -10,6 +10,22 @@
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
+import org.mockito.Mock;
+import org.junit.Ignore;
+import java.io.IOException;
```
Unused imports, we should remove.

bzamora020 replied:
Done

WadeVaresio commented on src/test/java/edu/ucsb/ucsbcslas/controllers/TutorAssignmentControllerTests.java:
```diff
@@ -10,6 +10,22 @@
+import org.springframework.mock.web.MockMultipartFile;
+import org.springframework.web.context.WebApplicationContext;
+import org.springframework.test.web.servlet.setup.MockMvcBuilders;
+import org.springframework.web.servlet.View;
```
Unused imports, we should remove.

bzamora020 replied:
done

WadeVaresio commented on src/test/java/edu/ucsb/ucsbcslas/controllers/TutorAssignmentControllerTests.java:
```diff
@@ -10,6 +10,22 @@
 import org.springframework.test.web.servlet.MvcResult;
 
+// Imports from BZ when implementing uploading CSV functionality to Tutor Assignment
+import org.springframework.web.multipart.MultipartFile;
```
Unused imports, we should remove.

bzamora020 replied:
Done

WadeVaresio reviewed (changes requested):
Overall looks good, I have some small requested changes.

Secondly, would it be possible to display the "Upload CSV " Button when no Tutor Assignments exist, essentially whenever the "New Tutor Assignment" Button is displayed?

bzamora020 commented:
> Overall looks good, I have some small requested changes.
> 
> Secondly, would it be possible to display the "Upload CSV " Button when no Tutor Assignments exist, essentially whenever the "New Tutor Assignment" Button is displayed?

I added the showing the upload CSV part to it, deleted the comments, removed the not used imports, made it check both on front and backend that the user is an authorized user to make an upload CSV request. 

When I introduced the way to check for an authorized user to make an upload csv request, the coverage of a line that was previously covered is not covered. Is it possible to make a PR after to fix that?

WadeVaresio commented on src/main/java/edu/ucsb/ucsbcslas/controllers/TutorAssignmentController.java:
```diff
@@ -221,55 +230,114 @@
+                throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Malformed CSV", e);
+            }
+        } else {
+            return null;
```
I think instead of returning null here, we want to return a BAD_REQUEST similar to line 336.

WadeVaresio commented on src/main/java/edu/ucsb/ucsbcslas/controllers/TutorAssignmentController.java:
```diff
@@ -221,55 +230,114 @@
-      tutorAssignmentRepository.save(newAssignment);
-      String body = mapper.writeValueAsString(newAssignment);
-      return ResponseEntity.ok().body(body);
+            } catch (RuntimeException e) {
+                logger.error("Error Uploading CSV", e);
+                throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Malformed CSV", e);
```
To cover lines 334-336, a test needs to be written that deliberately calls this endpoint with incorrect user permissions. Thus, we essentially mock the authControllerAdvice to return that the current user is not an admin.

JacquelineMai reviewed (changes requested):
I'm less concerned with the missing coverage for the frontend files, since I think we know that it's hard to test toasts. Coverage should be added on this PR for the TutorAssignmentController, though.


For the RuntimeException catch, you can follow this test as an example, which mocks a parse failure: https://github.com/ucsb-cs156-s21/proj-ucsb-cs-las/blob/daa26ff9b0374aa7a016c1c29500d0d019984d9e/src/test/java/edu/ucsb/ucsbcslas/controllers/CourseControllerTests.java#L566

For line 339 you should add a new test where the user is not authenticated

WadeVaresio reviewed (approved):
LGTM, all requested changes have been made. Great job Bryan!

JacquelineMai reviewed (changes requested):
This seems to have a bug on Heroku, where the Tutor Assignments page doesn't load at all. I've already discussed this with the team, but leaving this review so this doesn't get pre-emptively merged.

WadeVaresio commented:
The bug appears to be from white space around entity field values, when there should not be whitespace. The fix is to trim white space when setting each value in `TutorAssignmentModel.java`, as the parse does not use the parametrized constructor rather uses the set methods. Thus the parametrized constructor can also be removed from `TutorAssignmentModel.java`.

WadeVaresio commented on javascript/src/main/pages/TutorAssignment/TutorAssignments.js:
```diff
@@ -4,31 +4,87 @@ import { Button } from "react-bootstrap";
+  const {
+    data: tutorAssignmentList,
+    error,
+    mutate: mutateTutorAssignments, mutateCourses,
```
We can remove the mutateCourse reference here.

bzamora020 replied:
done

WadeVaresio commented on javascript/src/main/pages/TutorAssignment/TutorAssignments.js:
```diff
@@ -39,52 +95,81 @@ const TutorAssignment = () => {
     getToken, mutateCourses, onDeleteTutorError
```
Instead of `mutateCourse` we should replace this with `mutateTutorAssignments`, this should fix the bug observed where there is a noticeable delay between deleting a tutor assignment, and the results being updated on the table.

bzamora020 replied:
Fixed bug, it did help to remove the delay

WadeVaresio reviewed (changes requested):
LGTM, we are almost ready to merge once the small requested changes are made, and a team member reviews the PR.

bzamora020 commented on javascript/src/main/pages/TutorAssignment/TutorAssignments.js:
```diff
@@ -4,31 +4,87 @@ import { Button } from "react-bootstrap";
+      >
+        Required Columns: Course Name, Quarter(In numeric format), Instructor First Name,
+        Instructor Last Name, Instructor Email, Tutor First Name, Tutor Name,
+        Tutor Email, Assignment Type.Ex: CMPSC 48, 20201, Joe, Gaucho, joegaucho
```
Fixed

WadeVaresio reviewed (approved):
LGTM, great job! Small gap in front end test coverage has been okayed to merge, per Bryan speaking to Professor Conrad.

dorinori reviewed (approved):
LGTM
